### <b> Pandas to work on dataframe

In [1]:
import pandas as pd

In [3]:
df = pd.read_csv('anneal.data')

In [ ]:
df.shape

In [ ]:
df.info()

In [ ]:
df.head()

In [ ]:
df

In [ ]:
# df.S.value_counts()['?']
df.isin(['?']).sum(axis=0)

In [ ]:
df.columns

In [ ]:
trainW = df

### Make a deep copy because '=' is not sufficient in our example

In [ ]:
hello = 5
hi = hello
hi

In [ ]:
hi = 4
hello

In [ ]:
hi

In [ ]:
hello = 3
hi

In [ ]:
hi = hello
hi

In [ ]:
hello = 5
hi

In [ ]:
hi == hello

<img src="https://media.geeksforgeeks.org/wp-content/uploads/deep-copy.jpg"></img>

<img src="https://media.geeksforgeeks.org/wp-content/uploads/shallow-copy.jpg">

### Learn more about copy [here](https://docs.python.org/3/library/copy.html#:~:text=A%20shallow%20copy%20constructs%20a,objects%20found%20in%20the%20original) <font color='red'><b>This is official</font><font color=green> python 3.9</font> doc

In [5]:
import copy as cp

In [6]:
trainW = cp.copy(df)
trainW

?  C  A  08  00 ?.1  S ?.2  000 ?.3  ... ?.22 ?.23   COIL  0.700  0610.0  \
0    ?  C  R   0   0   ?  S   2    0   ?  ...    ?    ?   COIL  3.200   610.0   
1    ?  C  R   0   0   ?  S   2    0   ?  ...    ?    ?  SHEET  0.700  1300.0   
2    ?  C  A   0  60   T  ?   ?    0   ?  ...    ?    ?   COIL  2.801   385.1   
3    ?  C  A   0  60   T  ?   ?    0   ?  ...    ?    ?  SHEET  0.801   255.0   
4    ?  C  A   0  45   ?  S   ?    0   ?  ...    ?    ?   COIL  1.600   610.0   
..  .. .. ..  ..  ..  .. ..  ..  ...  ..  ...  ...  ...    ...    ...     ...   
792  ?  C  A   0  50   T  ?   ?    0   ?  ...    ?    ?   COIL  1.001    50.0   
793  ?  C  ?   0   0   ?  S   1    0   ?  ...    ?    ?  SHEET  0.699  1300.0   
794  ?  C  A   0   0   ?  S   2    0   ?  ...    ?    ?   COIL  0.400   609.9   
795  ?  C  A   0   0   ?  S   2    0   ?  ...    ?    ?  SHEET  3.200   610.0   
796  ?  C  R   0   0   ?  S   2    0   ?  ...    ?    ?  SHEET  1.599  1500.0   

     0000 ?.24 0000.1 ?.25  3  
0       0    ?      0    ?  3  
1     762    ?      0    ?  3  
2       0    ?      0    ?  3  
3     269    ?      0    ?  3  
4       0    ?      0    ?  3  
..    ...  ...    ...  ... ..  
792     0    Y      0    ?  3  
793  4880    ?      0    ?  3  
794     0    ?      0    ?  3  
795  4880    ?      0    ?  3  
796  4170    ?      0    ?  2  

[797 rows x 39 columns]

In [7]:
testW = pd.read_csv('anneal.test')
testW

?  C  A  00  45 ?.1  S ?.2  000 ?.3  ... ?.22 ?.23   COIL  1.600  0610.0  \
0    ?  C  A   0   0   ?  S   3    0   N  ...    ?    ?   COIL  0.699   609.9   
1   ZS  C  A   0  85   T  ?   ?    0   ?  ...    ?    ?  SHEET  0.400   610.0   
2   ZS  C  A   0  50   T  ?   ?    0   ?  ...    ?    ?   COIL  0.400   610.0   
3    ?  C  A   0   0   ?  S   2    0   ?  ...    ?    ?   COIL  0.699  1320.0   
4   TN  C  ?   0   0   ?  A   1    0   ?  ...    ?    ?   COIL  0.800    50.0   
..  .. .. ..  ..  ..  .. ..  ..  ...  ..  ...  ...  ...    ...    ...     ...   
94   ?  C  R   0   0   ?  S   3    0   ?  ...    ?    ?  SHEET  1.599   610.0   
95   ?  C  R   0   0   ?  S   3    0   ?  ...    ?    ?  SHEET  1.601   830.0   
96   ?  C  V   0   0   ?  S   2    0   ?  ...    ?    ?  SHEET  1.599   150.0   
97   ?  C  A   0  85   T  ?   ?    0   ?  ...    ?    ?   COIL  0.400    20.0   
98   ?  C  A   0  85   T  ?   ?    0   ?  ...    ?    ?   COIL  4.000   610.0   

   0000 ?.24 0000.1 ?.25  3  
0     0    ?      0    ?  3  
1   762    ?      0    ?  U  
2     0    ?      0    ?  3  
3     0    ?      0    ?  3  
4     0    ?      0    ?  5  
..  ...  ...    ...  ... ..  
94  762    ?      0    ?  2  
95  880    ?      0    ?  2  
96  762    ?      0    ?  2  
97    0    ?      0    ?  U  
98    0    ?    500    ?  U  

[99 rows x 39 columns]

In [ ]:
trainW.shape

In [ ]:
testW.shape

### Columns will be added upon concatenation
#### trainW.shape = (797, 39)    
#### testW.shape = (99, 39)
#### W.shape (797+99, 39) i.e. (896, 39)
#### W.shape = (896, 39)

### Vim [Macro](https://vim.fandom.com/wiki/Macros) used
#### <B> Logic</b>
1> go to start of line 2> go to : and delete rest of line 3> insert at end of line , 4> normal mode 5> go to start of line, 6> down 1 line.

#### @ <b>qdf:Da,{esc}{down}

In [8]:
columns = ["family", "product-type", "steel", "carbon", "hardness", "temper_rolling",
"condition", "formability", "strength", "non-ageing", "surface-finish", "surface-quality",
"enamelability", "bc", "bf", "bt", "bw/me", "bl", "m", "chrom", "phos", "cbond", "marvi",
"exptl", "ferro", "corr", "blue/bright/varn/clean", "lustre", "jurofm", "s", "p", "shape",
"thick", "width", "len", "oil", "bore", "packing", "classes"]

In [9]:
W = pd.concat([trainW, testW], ignore_index=True)
W.shape

(896, 42)

In [ ]:
# W.info
W.columns

In [ ]:
W.columns = columns

In [ ]:
trainW.iloc[:, 35:-1]

In [ ]:
W.iloc[5:, -3:]

In [12]:
X = cp.copy(W.iloc[:,:-3])

In [11]:
X.iloc[:,-3:]

NameError: name 'X' is not defined

In [13]:
X.columns = columns
X.columns

Index(['family', 'product-type', 'steel', 'carbon', 'hardness',
       'temper_rolling', 'condition', 'formability', 'strength', 'non-ageing',
       'surface-finish', 'surface-quality', 'enamelability', 'bc', 'bf', 'bt',
       'bw/me', 'bl', 'm', 'chrom', 'phos', 'cbond', 'marvi', 'exptl', 'ferro',
       'corr', 'blue/bright/varn/clean', 'lustre', 'jurofm', 's', 'p', 'shape',
       'thick', 'width', 'len', 'oil', 'bore', 'packing', 'classes'],
      dtype='object')

In [ ]:
X

In [ ]:
X.isin(['?']).sum(axis=0)

#### [Remove columns with over 15% missing values](https://stackoverflow.com/questions/37921703/how-to-delete-columns-with-at-least-20-missing-values)

In [15]:
Xx = X.loc[:, X.isin(["?"]).sum()<len(X)*.20]
Xx.isin(['?']).sum(axis=0)

product-type     0
steel           86
carbon           0
hardness         0
strength         0
shape            0
thick            0
width            0
len              0
bore             0
classes          0
dtype: int64

In [17]:
Xx['surface_quality'] = X['surface-quality']

/home/julia/.local/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [18]:
Xx.loc[:,"surface_quality"]

0        E
1        E
2        G
3        G
4        D
      ... 
891    NaN
892    NaN
893    NaN
894    NaN
895    NaN
Name: surface_quality, Length: 896, dtype: object

In [25]:
# Xx = Xx.drop(columns=["surface_quality"])
Xx

product-type steel  carbon  hardness  strength  shape  thick   width  len  \
0              C     R     0.0         0         0   COIL  3.200   610.0    0   
1              C     R     0.0         0         0  SHEET  0.700  1300.0  762   
2              C     A     0.0        60         0   COIL  2.801   385.1    0   
3              C     A     0.0        60         0  SHEET  0.801   255.0  269   
4              C     A     0.0        45         0   COIL  1.600   610.0    0   
..           ...   ...     ...       ...       ...    ...    ...     ...  ...   
891            C     R     NaN         0         0  SHEET    NaN   610.0  762   
892            C     R     NaN         0         0  SHEET    NaN   830.0  880   
893            C     V     NaN         0         0  SHEET    NaN   150.0  762   
894            C     A     NaN         0         0   COIL    NaN    20.0    0   
895            C     A     NaN         0         0   COIL    NaN   610.0    0   

     bore classes  
0       0       3  
1       0       3  
2       0       3  
3       0       3  
4       0       3  
..    ...     ...  
891     0       2  
892     0       2  
893     0       2  
894     0       U  
895   500       U  

[896 rows x 11 columns]

In [29]:
Xx['surface_quality'] = X['surface-quality']

In [30]:
Xx.columns

Index(['product-type', 'steel', 'carbon', 'hardness', 'strength', 'shape',
       'thick', 'width', 'len', 'bore', 'classes', 'surface_quality'],
      dtype='object')

In [31]:
Xx = Xx[['product-type', 'steel', 'carbon', 'hardness', 'strength', 'shape',
       'thick', 'width', 'len', 'bore', 'surface_quality', 'classes']]
Xx.columns

Index(['product-type', 'steel', 'carbon', 'hardness', 'strength', 'shape',
       'thick', 'width', 'len', 'bore', 'surface_quality', 'classes'],
      dtype='object')

In [ ]:
Xx.loc[:,"temper_rolling"].isin(['?']).sum()

In [34]:
Xx2 = X.loc[:, X.isin(["?"]).sum()<len(X)*.20]
Xx2.isin(['?']).sum(axis=0)

product-type     0
steel           86
carbon           0
hardness         0
strength         0
shape            0
thick            0
width            0
len              0
bore             0
classes          0
dtype: int64

In [ ]:
len(Xx2.columns)

In [35]:
Xx3 = Xx2.iloc[:,:-1]
y = Xx2.iloc[:,-1]

#### Check what is in column 'steel' that has 86 ? so that we can replace ? with appropriate value, if column is continuous data then mean or median would be appropriate and for cetagorical data mod would be best strategy

In [ ]:
Xx3.loc[:,'steel']

#### so mod it is, since the column has cetagories or classes

In [ ]:
Xx3

In [ ]:
X

In [ ]:
Xx3.isnull().sum()

In [ ]:
Xx3.loc[:,["carbon", "thick"]]

In [ ]:
from collections import Counter
Counter(Xx3.loc[:,["carbon", "thick"]])

#### Let's delete columns which have less then 2 categories since they will not help classify the target 

In [ ]:
Counter(Xx3)

In [ ]:
# Xx3.groupby(Xx3.columns)
len(Xx3.columns)

In [ ]:
len(Xx2.columns)

In [ ]:
Xx3

In [ ]:
Xx2

In [ ]:
X.iloc[:,-5:]

In [ ]:
X.loc[:,["temper_rolling", "surface-quality", "enamelability","formability"]].isin(["?"]).sum()

In [ ]:
X.groupby("surface-quality").size()

#### ok made a mistake this column "surface-quality" has some information about it, I should have not deleted it, but I am too lazy to fix it now

In [ ]:
# X.groupby("carbon").size()
X.groupby("steel").size()

In [ ]:
import numpy as np

In [ ]:
Xx3.loc[Xx3.steel.isin(['?']), :]
# XXX = Xx3.loc[Xx3.steel.isin(['?']), :] = np.NaN # potentially dangerous since it will turn everything to np.nan

In [ ]:
# Xx3.thick.isnull().sum()
# Xx3.steel.isnull().sum()
Xx3.carbon.isnull().sum()

#### impute the missing values, the column "steel" has categorical data which I am going to replace for "?" with most frequent occourance

In [ ]:
from sklearn.impute import SimpleImputer as simp

In [ ]:
mf = simp(missing_values="?", strategy='most_frequent')
mean = simp(missing_values=np.nan, strategy='mean')

In [ ]:
XX3 =cp.copy(Xx3)

In [36]:
# mf.fit(XX3.loc[:,["steel"]])
# XX3.loc[:,"steel"]
Xx3.loc[Xx3.steel.isnull(), :]

Empty DataFrame
Columns: [product-type, steel, carbon, hardness, strength, shape, thick, width, len, bore]
Index: []

In [2]:
217/863

0.2514484356894554